### PR TITLE
Rh6 request uac

### PR DESCRIPTION
**Pull request by Paul-Joel**
# Motivation and Context
Delivery of code for RH6 - Request a new access code

# What has changed
Design pattern upped to 13.3.0
New journey introduced, starting from homepage link to 'Request a new code'
Journey connects to address index postcode endpoint, and 3 new RHSVC endpoints (already delivered)

# How to test?
Follow link from homepage /start/, and follow instructions.
Running locally, Address Index will connect to the Exeter subset, so only Exeter postcodes will return addresses. 'EX2 6GA' and 'EX1 1LL' will return addresses

# Links
None

# Screenshots (if appropriate):
None

### CODE REVIEW

File: tests/unit/test_handlers.py [hunk 1]
```diff
@@ -567,3 +568,164 @@ def test_validate_case_empty(self):
...
+
+            with self.assertLogs('respondent-home', 'INFO') as cm:
+                response = await self.client.request("POST", self.post_requestcode, data=self.request_code_form_data_valid)
+                self.assertLogLine(cm, "Valid postcode")
+
+                self.assertEqual(response.status, 200)
+                self.assertIn('We cannot find your address', str(await response.content.read()))
+
+    @unittest_run_loop
+    async def test_post_request_code_select_address(self):
```
Comment (MartinHumphrey) [inline tests/unit/test_handlers.py]: Duplicate of test_post_request_access_code_good_postcode with my comment to look for a potential selection option in the content returned added. Could keep both but exercise exactly the same code with same result?

File: app/handlers.py [hunk 2]
```diff
@@ -59,6 +61,22 @@ def _rhsvc_url_surveylaunched(self):
     def _webchat_service_url(self):
         return self._request.app['WEBCHAT_SVC_URL']
 
+    @property
+    def _ai_url_postcode(self):
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/handlers.py]: Actioned - Request specific properties and methods moved to RequestCommon block

File: app/templates/request-code-household.html [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+{% block main %}
+
+    {% if messages_dict %}
+        {% include "partials/messages.html" with context %}
+    {% endif %}
+
+    {% call onsQuestion({
+        "title": "What is your postcode?",
+        "description": "<p>To text you a new code we need to know what address you are answering for.</p>"
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/templates/request-code-household.html]: Actioned: Updated to your text - this is still not agreed, but your text reads better, so implemented until final text agreed

File: app/handlers.py [hunk 4]
```diff
@@ -394,3 +431,244 @@ def validate_form(self, data):
...
+                 "id": singleAddress['uprn']})
+
+        address_content = {'postcode': postcode,
+                           'addresses': address_options,
+                           'display_region': display_region,
+                           'total_matches': postcode_return['response']['total']}
+
+        return address_content
+
+    postcode_validation_pattern = re.compile(r'^([A-PR-UWYZa-pr-uwyz]([0-9]{1,2}|([A-HK-Ya-hk-y][0-9]|[A-HK-Ya-hk-y][0-9]([0-9]|[ABEHMNPRV-Yabehmnprv-y]))|[0-9][A-HJKS-UWa-hjks-uw])\ {0,1}[0-9][ABD-HJLNP-UW-Zabd-hjlnp-uw-z]{2}|([Gg][Ii][Rr]\ 0[Aa][Aa])|([Ss][Aa][Nn]\ {0,1}[Tt][Aa]1)|([Bb][Ff][Pp][Oo]\ {0,1}([Cc]\/[Oo]\ )?[0-9]{1,4})|(([Aa][Ss][Cc][Nn]|[Bb][Bb][Nn][Dd]|[BFSbfs][Ii][Qq][Qq]|[Pp][Cc][Rr][Nn]|[Ss][Tt][Hh][Ll]|[Tt][Dd][Cc][Uu]|[Tt][Kk][Cc][Aa])\ {0,1}1[Zz][Zz]))$')  # NOQA
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/handlers.py]: Actioned: Updated to CC regex, and match done against uppercase

File: app/templates/request-code-select-address.html [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        <p>If this is incorrect you can <a href="{{ url('RequestCode:get') }}">enter the address</a> again or <a href="#">contact us</a> for help.</p>
+
+    {% else: %}
+
+        {% call onsQuestion({
+            "title": "Select your address",
+            "description": question_description
+        }) %}
+
+            {% if 'address' in field_messages_dict %}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/templates/request-code-select-address.html]: Actioned: Resolved with next issue - oversight resulting from design change

File: app/handlers.py [hunk 6]
```diff
@@ -394,3 +431,244 @@ def validate_form(self, data):
...
+        address_content = await self.get_postcode_return(attributes["postcode"], attributes["display_region"])
+
+        return address_content
+
+    @aiohttp_jinja2.template('request-code-select-address.html')
+    async def post(self, request):
+        await self.get_check_attributes(request)
+        data = await request.post()
+
+        form_return = ast.literal_eval(data["request-address-select"])
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/handlers.py]: Actioned: Now checks for value and throws flash message and error if empty

File: app/handlers.py [hunk 7]
```diff
@@ -394,3 +431,244 @@ def validate_form(self, data):
...
+
+        session = await get_session(request)
+        session["attributes"]["address"] = form_return["address"]
+        session["attributes"]["uprn"] = form_return["uprn"]
+        session.changed()
+        logger.warn("Session updated", client_ip=self._client_ip)
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/handlers.py]: Actioned: Copy mistake, changed to 'info'

File: app/handlers.py [hunk 8]
```diff
@@ -394,3 +431,244 @@ def validate_form(self, data):
...
+            return attributes
+
+        if address_confirmation == 'yes':
+
+            session = await get_session(request)
+            uprn = session["attributes"]['uprn']
+
+            try:
+                uprn_return = await self.get_cases_by_uprn(uprn)
+                session["attributes"]["case_id"] = uprn_return[0]["id"]
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/handlers.py]: Actioned: Comment added as reminder that this needs revisiting post 2019

File: tests/unit/test_handlers.py [hunk 9]
```diff
@@ -567,3 +568,164 @@ def test_validate_case_empty(self):
...
+    async def test_post_request_access_code_good_postcode(self):
+
+        with mock.patch('app.handlers.RequestCodeCommon.get_ai_postcode') as mocked_get_ai_postcode:
+            mocked_get_ai_postcode.return_value = self.ai_postcode_results
+
+            with self.assertLogs('respondent-home', 'INFO') as cm:
+                response = await self.client.request("POST", self.post_requestcode, data=self.request_code_form_data_valid)
+            self.assertLogLine(cm, "Valid postcode")
+
+            self.assertEqual(response.status, 200)
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/unit/test_handlers.py]: Actioned: Combined 3 existing tests into one, so check for 'Valid postcode' log line, 200 status, 'Select an address' text and first address option all in one test

File: tests/unit/test_handlers.py [hunk 10]
```diff
@@ -567,3 +568,164 @@ def test_validate_case_empty(self):
...
+
+    @unittest_run_loop
+    async def test_post_request_access_code_found(self):
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/unit/test_handlers.py]: Actioned: Test removed (see above)

File: tests/unit/test_handlers.py (hunk 1 above)
Comment (Paul-Joel) [reply tests/unit/test_handlers.py]: Actioned: Test removed/combined (see above)

**Pull request merged**